### DISCUSSION TITLE
Use other Entity Repository in Hooks

### DISCUSSION THREAD

**Discussion by HunterJS-bit** [mikro-orm/mikro-orm#1666] [Q&A] [answered]
Hi guys, I would like to insert data into other repository or in Other model, when hooks has happened

For example I have model User:

```

import {
    AfterUpdate,
    Cascade,
    Collection,
    Entity,
    MikroORM,
    ManyToOne,
    OneToMany,
    PrimaryKey,
    Property
} from '@mikro-orm/core';
import { UserRepository } from '../modules/user/user.repository.ts';

@Entity({ tableName: 'users', customRepository: () => AssetTypeAttributeLabelRepository })
export class User {
    constructor(data?) {
        if (data) {
            Object.keys(data).forEach((key) => {
                this[key] = data[key];
            });
        }
    }

    @PrimaryKey()
    id: number;

    @Property()
    phone = false;

    // @ts-ignore
    @AfterUpdate()
    async doStuff(evt) {
        const { entity, originalEntity } = evt.changeSet;
        if (entity.phone !== originalEntity.phone) {
            console.log('about to do some stuff hererer');
            // how can I call here Other repository or other service to insert data in other model

        }
    }

}
```

And I have **AfterUpdate Hook**, how would I call from there Other model or other service or repository to Insert Data into other Model, for example in model Contacts ?

**Comment by B4nan** [comment, marked as answer]
You can either use `onFlush` event if you want to add the new entity to the same batch (same transaction) as the parent one is persisted. 

https://mikro-orm.io/docs/lifecycle-hooks/#using-onflush-event

Or you can use `em.nativeInsert` to do the insert query without having side effects (without touching the identity map). 

Flushing from inside hooks is forbidden.

**Reply by HunterJS-bit** [reply]
Great thank you, will look a bit more into the docs :)

**Comment by kennethsato** [comment]
I have a question that I believe is closely related to this discussion.

I (now) understand that I should not flush inside a lifecycle hook:
https://mikro-orm.io/docs/events#limitations-of-lifecycle-hooks

But I've noticed that after coding some hooks with a couple of flushs and persists.

I have a AfterUpdate hook that logs any entity changes creating log entries on a database, like this:
```
  @AfterUpdate()
  async afterUpdate(eventArgs: EventArgs<Chamada>): Promise<void> {
    if (eventArgs.changeSet) {
      const payload = Object.keys(eventArgs.changeSet.payload);
      const usuario_id = eventArgs.em.getFilterParams('chamada').usuario_id;
      const chamada = eventArgs.changeSet.entity as Chamada;
      const prevValues = eventArgs.changeSet.originalEntity as Chamada;
      const createLog = container.resolve(CreateLogService);

      console.log({ payload });
      payload.forEach(async field => {
        const mensagem = `Campo ${field.toUpperCase()} alterado de ${
          prevValues[field as keyof Chamada]
        } para ${chamada[field as keyof Chamada]}`;

        await createLog.execute({
          mensagem,
          chamada_id: chamada.id,
          usuario_id,
        });
      });
    }
  }
```
And after upgrading to v6, I got the following error:

> You cannot call em.flush() from inside lifecycle hook handlers

But I have something similiar in another entity, but it does not have a forEach loop. It just writes a single log entry.
And I don't get an error there.

Would that be expected?
Or any flush/persist could cause unexpected consequences, once it is not recommended?

Should I remove the AfterUpdate hook and create those logs on our Service and manually compare each field to know each field was updated?

**Reply by B4nan** [reply]
Your `forEach` call ignores the promises it fires inside it, the code is wrong like this, you need a `for of` loop with `await`. And yes, you cannot flush inside a hook, so if your `createLog.execute` is flushing, that code wont work, you need `onFlush` hook as described in the docs.

**Reply by B4nan** [reply]
Alternatively, you can create a new EM fork inside the hook and flush that. Or just do `em.insert` which wont touch the context anyhow.

**Reply by kennethsato** [reply]
Replacing forEach to a for... of worked! Thanks @B4nan 
As long as my createLog service creates a new EntityManager, I can flush and persist data to my database on my lifecycle hook?

Besides, I was trying to use onFlush as an alternative, but I was not sure if I was coding it correctly.
I've created the subscriber class based on my Chamada Entity:
```
import { EventSubscriber, FlushEventArgs } from '@mikro-orm/core';
import Chamada from '../entities/Chamada';

export class ChamadaSubscriber implements EventSubscriber<Chamada> {
  async onFlush(args: FlushEventArgs): Promise<void> {
    console.log(args.uow.getChangeSets());
  }
}
```
And registered it on my service:
```
this.em.getEventManager().registerSubscriber(new ChamadaSubscriber());
```
It was firing and I was able to get the payloads

Was that correct?

**Reply by B4nan** [reply]
`onFlush` is not bound to an entity, its fired once per flush call. Check the docs, there is an example on how to use it.

**Reply by kennethsato** [reply]
Sorry.
I've read it, but I was not fully understanding it.

```
import {
  ChangeSetType,
  EventSubscriber,
  FlushEventArgs,
} from '@mikro-orm/core';
import Chamada from '../entities/Chamada';

export class ChamadaSubscriber implements EventSubscriber {
  async onFlush(args: FlushEventArgs): Promise<void> {
    console.log(args.uow.getChangeSets());
    const changeSets = args.uow.getChangeSets();
    const cs = changeSets.find(
      cs => cs.type === ChangeSetType.UPDATE && cs.entity instanceof Chamada,
    );
    if (cs) {
      // Code for Chamada @AfterUpdate
    }
  }
}
```

So should it be something like this?

**Reply by B4nan** [reply]
I'd say you should just go with `em.insert` if all you need is to run an insert query. If its an entity graph (so several connected entities), using a fork with a flush would work fine too, unless you want to run this inside the same transaction. Note that the `em.insert` approach will run things in the same transaction too, so if all you need is a one or two insert statements, no need to deal with the onFlush event.

But yes, something like this would be correct to detect a change of existing entity by its type, and inside the if block you would do `args.uow.computeChangeSet(newEntity);` on whatever new entity you want to create in this context.
